crates/server/src/middleware/jwt.rs

```diff
@@ -0,0 +1,159 @@
+use core::convert::Infallible;
+use core::fmt::{self, Display, Formatter};
+use core::task::{Context, Poll};
+use std::error::Error;
+
+use axum::body::Body;
+use axum::extract::Request;
+use axum::http::{HeaderMap, StatusCode};
+use axum::response::{IntoResponse, Response};
+use calimero_store::Store;
+use jsonwebtoken::{decode, DecodingKey, Validation};
+use libp2p::futures::future::BoxFuture;
+use tower::{Layer, Service};
+use tracing::debug;
+
+use crate::admin::storage::jwt_secret::get_jwt_secret;
+use crate::admin::utils::jwt::Claims;
+
+#[derive(Clone)]
+pub struct JwtLayer {
+    store: Store,
+}
+
+impl JwtLayer {
+    pub const fn new(store: Store) -> Self {
+        Self { store }
+    }
+}
+
+impl<S> Layer<S> for JwtLayer {
+    type Service = JwtMiddleware<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        JwtMiddleware {
+            inner,
+            store: self.store.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct JwtMiddleware<S> {
+    inner: S,
+    store: Store,
+}
+
+impl<S> Service<Request<Body>> for JwtMiddleware<S>
+where
+    S: Service<Request<Body>, Response = Response<Body>, Error = Infallible> + Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = Infallible;
+    // `BoxFuture` is a type alias for `Pin<Box<dyn Future + Send + 'a>>`
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request<Body>) -> Self::Future {
+        // todo! experiment with Interior<Store>: WriteLayer<Interior>
+        let result = auth(request.headers(), &self.store.clone());
+
+        if let Err(err) = result {
+            let error_response = err.into_response();
+            return Box::pin(async move { Ok(error_response) });
+        }
+
+        let future = self.inner.call(request);
+
+        Box::pin(async move {
+            let response: Response = future.await?;
+            Ok(response)
+        })
+    }
+}
+
+#[derive(Debug)]
+struct JwtHeader {
+    token: String,
+}
+
+pub fn auth(headers: &HeaderMap, store: &Store) -> Result<(), UnauthorizedError<'static>> {
+    let jwt_header = get_jwt_token_from_headers(headers).map_err(|e| {
+        debug!("Failed to extract authentication headers {}", e);
+        UnauthorizedError::new("Failed to extract authentication headers.")
+    })?;
+
+    let jwt_secret = match get_jwt_secret(&store.clone()) {
+        Ok(Some(secret)) => secret.jwt_secret().to_vec(),
+        Ok(None) => {
+            return Err(UnauthorizedError::new("JWT secret not found."));
+        }
+        Err(_) => {
+            return Err(UnauthorizedError::new("Failed to fetch JWT secret."));
+        }
+    };
+
+    let _token_data = decode::<Claims>(
```

Comment:
we still need to ensure it hasn't expired, yeah?

Reply:
Added check if expired or not